AdaptiveRoads/UI/QuayRoads/QuayRoadsPanel.cs

```diff
@@ -0,0 +1,231 @@
+            addSectionButton.eventClicked += (_, _) => {
+                profile_ = profile_.Expand(profile_.Length + 1, (_) => ProfileSection.Default());
+                Refresh();
+            };
+
+            var removeSectionButton = buttonsPanel.AddUIComponent<UIButtonExt>();
+            removeSectionButton.text = "remove section";
+            removeSectionButton.eventClicked += (_, _) => {
+                profile_ = profile_.Shrink(profile_.Length - 1, (_, _) => { });
+                Refresh();
+            };
+
+            var deleteButton = buttonsPanel.AddUIComponent<UIButtonExt>();
+            deleteButton.text = "delete";
+            deleteButton.eventClicked += (_, _) => {
+                profile_ = null;
+                Close();
+            };
+
+            if(profile_ is null) {
+                profile_ = new ProfileSection[] { };
+            }
+
+            
+            Refresh();
+        }
+        public void Refresh() {
+            if (propertiesTable_ is not null) {
+                RemoveUIComponent(propertiesTable_);
+                DestroyImmediate(propertiesTable_.gameObject);
+            }
+
+            propertiesTable_ = AddUIComponent<UITable>();
+            makeFirstColumn_();
+            for (int sectionIndex = 0; sectionIndex < profile_.Length; sectionIndex++) {
+                makeEditableColumn_(sectionIndex);
+            }
+        }
+        public override void Close() {
+            instances_.Remove(netInfo_);
+            base.Close();
+        }
+
+        public static void CloseIfOpen(NetInfo netInfo) {
+            if (instances_.ContainsKey(netInfo)) {
+                instances_[netInfo].Close();
+            }
+        }
+        public static QuayRoadsPanel GetOrOpen(NetInfo netInfo, RoadEditorPanel parentPanel) {
+            if (instances_.ContainsKey(netInfo)) {
+                return instances_[netInfo];
+            } else {
+                var newPanel = Create();
+                newPanel.netInfo_ = netInfo;
+                newPanel.parentPanel_ = parentPanel;
+                instances_[netInfo] = newPanel;
+                return newPanel;
+            }
+        }
+        public static void CloseAll() {
+            foreach(var instance in instances_) {
+                instance.Value.Close();
+            }
+        }
+
+        private struct PropertyGroup {
+            public string GroupName;
+
+            public PropertyGroup(string groupName) : this() {
+                GroupName = groupName;
+                Properties = new();
+            }
+
+            public List<Property> Properties;
+        }
+        private struct Property {
+            public FieldInfo fieldInfo;
+            public IConvertible? flag;
+
+            public Property(FieldInfo fieldInfo, IConvertible? flag = null) {
+                this.fieldInfo = fieldInfo;
+                this.flag = flag;
+            }
+        }
+
+        // TODO: precompute this. When?
```

Comment:
// TODO: precompute this. When?

Reply:
the part of the code that is in response to User click does not need to be fast.
if it was done every frame that was a different story.